module/zfs/zio_inject.c

```diff
@@ -1122,6 +1150,52 @@ zio_inject_list_next(int *id, char *name, size_t buflen,
 	return (ret);
 }
 
+/*
+ * Waits on injection events (handler added, removed, or event injected).
+ * Callers set *state to 0 initially and should pass the updated value back to
+ * successive calls to ensure that no events are not lost between wait calls.
+ * If timeout != NULL, waits
+ * Returns 0 if an event occurred since *state was updated, EINTR if a signal is
+ * caught, or ETIMEDOUT if no event occurred within the timeout.
+ */
+int
+zio_inject_wait(uint64_t *state, hrtime_t timeout)
+{
+	int error;
+	int rc;
+
+	mutex_enter(&inject_event_mtx);
+	if (timeout > 0 && *state > inject_event_count) {
+		rc = cv_timedwait_sig_hires(&inject_event_cv, &inject_event_mtx,
+		    timeout, USEC2NSEC(1), 0);
```

Comment:
I would expect to see a loop somewhere here.  Also I am not sure where "Waits forever if timeout is omitted" is implemented.  I would expect to see `cv_wait_sig()` there.